rules/private/diff_test_private.bzl

```diff
@@ -0,0 +1,170 @@
+            outputs = [test_script],
+            executable = "cmd.exe",
+            arguments = ["/C", diff_script.path.replace("/", "\\")],
+            env = {
+                "DIFF_TEST_SRC1": ctx.file.file1.path.replace("/", "\\"),
+                "DIFF_TEST_SRC2": ctx.file.file2.path.replace("/", "\\"),
+                "DIFF_TEST_OUT": test_script.path.replace("/", "\\"),
+            },
+            progress_message = "Comparing files",
+            mnemonic = "CompareFiles",
+        )
+    else:
+        test_script = ctx.actions.declare_file(ctx.label.name + "-test.sh")
+
+        # Source of the shell command that generates 'test_script'.
+        #
+        # This command runs at build time and compares the two files using
+        # "diff". Then it writes a shell script ('test_script') which is the
+        # actual script executed by "bazel test".
+        #
+        # If the files are the same and were expected to be so, or if they are
+        # different and were expected to be so, then the command writes
+        # 'test_script' to do nothing and exit with zero. Otherwise the command
+        # writes 'test_script' to print an error and exit with 1.
+        #
+        # If one of the files was not a regular file (or it was a symlink whose
+        # target is not a regular file) then this shell command fails.
+        command = """set -euo pipefail
+if ! test -f "$1"; then echo "File not found: $1" >&2 ; exit 1; fi
+if ! test -f "$2"; then echo "File not found: $2" >&2 ; exit 1; fi
+if diff -q "$1" "$2" >/dev/null 2>&1; then echo -e {was_same} > "$3"; else echo -e {was_different} > "$3"; fi
+chmod +x "$3"
+"""
+
+        # Contents of 'test_script' depending on diff result and expectation.
+        if ctx.attr.expect_same:
+            was_same = "'#!/bin/sh\nexit 0'"
+            was_different = "\"#!/bin/sh\necho 'Expected \'$1\' and \'$2\' to be the same' >&2\nexit 1\""
+        else:
+            was_same = "\"#!/bin/sh\necho 'Expected \'$1\' and \'$2\' to be different' >&2\nexit 1\""
+            was_different = "'#!/bin/sh\nexit 0'"
+
+        # Run the shell command to generate 'test_script'.
+        ctx.actions.run_shell(
+            inputs = [ctx.file.file1, ctx.file.file2],
+            outputs = [test_script],
+            command = command.format(was_same = was_same, was_different = was_different),
+            arguments = [ctx.file.file1.path, ctx.file.file2.path, test_script.path],
+            progress_message = "Comparing files",
+            mnemonic = "CompareFiles",
+        )
+
+    return DefaultInfo(
+        executable = test_script,
+        files = depset(direct = [test_script]),
+        runfiles = ctx.runfiles(files = [test_script]),
+    )
+
+_diff_test = rule(
+    attrs = {
+        "file1": attr.label(mandatory = True, allow_single_file = True),
+        "file2": attr.label(mandatory = True, allow_single_file = True),
+        "expect_same": attr.bool(default = True),
```

Comment:
Do you actually need the `expect_same` attribute?
I don't see a good use-case for that. I'd suggest to keep the rule even simpler and remove it.